internal/controller/driver_controller.go

```diff
@@ -17,46 +17,1211 @@ limitations under the License.
+			if err := r.Delete(r.ctx, existingCsiDriver); err != nil {
+				log.Error(err, "Failed to delete existing CSI Driver resource")
+				return err
+			}
+			log.Info("CSI Driver resource deleted successfully")
+		} else {
+			log.Info("CSI Driver resource does not exist")
+		}
+
+		if err := r.Create(r.ctx, desiredCsiDriver); err != nil {
+			log.Error(err, "Failed to create a CSI Driver resource")
+			return err
+		}
+
+		log.Info("CSI Driver resource created successfully")
+	} else {
+		log.Info("CSI Driver resource already meets desired state")
+	}
+
+	return nil
+}
+
+func (r *driverReconcile) reconcileControllerPluginDeployment() error {
+	deploy := &appsv1.Deployment{}
+	deploy.Name = r.generateName("ctrlplugin")
+	deploy.Namespace = r.driver.Namespace
+
+	log := r.log.WithValues("deploymentName", deploy.Name)
+	log.Info("Reconciling controller plugin deployment")
+
+	opResult, err := ctrlutil.CreateOrUpdate(r.ctx, r.Client, deploy, func() error {
+		if err := ctrlutil.SetControllerReference(&r.driver, deploy, r.Scheme); err != nil {
+			log.Error(err, "Failed setting an owner reference on deployment")
+			return err
+		}
+
+		appName := deploy.Name
+		leaderElectionSpec := cmp.Or(r.driver.Spec.LeaderElection, &defaultLeaderElection)
+		pluginSpec := cmp.Or(r.driver.Spec.ControllerPlugin, &csiv1a1.ControllerPluginSpec{})
+		serviceAccountName := cmp.Or(
+			ptr.Deref(pluginSpec.ServiceAccountName, ""),
+			fmt.Sprintf("csi-%s-ctrlplugin-sa", r.driverType),
+		)
+		imagePullPolicy := cmp.Or(pluginSpec.ImagePullPolicy, corev1.PullIfNotPresent)
+		grpcTimeout := cmp.Or(r.driver.Spec.GRpcTimeout, defaultGRrpcTimeout)
+		logLevel := ptr.Deref(r.driver.Spec.Log, csiv1a1.LogSpec{}).LogLevel
+		forceKernelClient := r.isCephFsDriver() && r.driver.Spec.CephFsClientType == csiv1a1.KernelCephFsClient
+
+		leaderElectionArgs := []string{
+			utils.LeaderElectionContainerArg,
+			utils.LeaderElectionNamespaceContainerArg(r.driver.Namespace),
+			utils.LeaderElectionLeaseDurationContainerArg(leaderElectionSpec.LeaseDuration),
+			utils.LeaderElectionRenewDeadlineContainerArg(leaderElectionSpec.RenewDeadline),
+			utils.LeaderElectionRetryPeriodContainerArg(leaderElectionSpec.RetryPeriod),
+		}
+
+		deploy.Spec = appsv1.DeploymentSpec{
+			Replicas: pluginSpec.Replicas,
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{"app": appName},
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: utils.Call(func() map[string]string {
+						podLabels := map[string]string{}
+						maps.Copy(podLabels, pluginSpec.Labels)
+						podLabels["app"] = appName
+						return podLabels
+					}),
+					Annotations: maps.Clone(pluginSpec.Annotations),
+				},
+				Spec: corev1.PodSpec{
+					ServiceAccountName: serviceAccountName,
+					PriorityClassName:  ptr.Deref(pluginSpec.PrioritylClassName, ""),
+					Affinity:           pluginSpec.Affinity,
```

Comment:
we need to set pod anti affinity to make sure we don't run two provisioner pod on same node https://github.com/rook/rook/blob/3ccdb53cb66e5d4151559ebf78cc18dab105cbe5/pkg/operator/ceph/csi/spec.go#L525

Reply:
done